Add NiFi Registry Deployment

jasonrogena opened the pull request:
Adds the NiFi Registry deployment [in build.openlmis.org](http://build.openlmis.org/job/NiFi-Registry-deploy-to-shared/).

jasonrogena committed:
OLMIS-5086: Add NiFi Registry Deployment

Since OpenLMIS needs a shared NiFi Registry instance, add deployment
scripts for it (to be used with Jenkins).

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
OLMIS-5086: Add Terraform Def For NiFi Reg Deploy

Since the shared NiFi Registry instance will be deployed on an AWS
instance, add the definitions for that setup in Terraform files.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
OLMIS-5086: Add Docker Install Auto Scripts

Since the NiFi Registry host will have to have the Docker Daemon
running, create Ansible scripts for installing Docker and Docker Compose
on that host.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

craigappl commented:
I added @joshzamor  as a reviewer. 1) I think it's a great opportunity for him to see how this was done and 2) I don't have the technical knowledge to assess the process.

craigappl commented:
I added @joshzamor  as a reviewer. 1) I think it's a great opportunity for him to see how this was done and 2) I don't have the technical knowledge to assess the process.

jasonrogena committed:
OLMIS-5231: Use Ansible Tasks For TLS Gen Process

Since Ansible is unable to generate enough entropy when running the
shell script for generating the TLS files in the Docker role, switch to
generating the TLS files using Ansible tasks.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena commented:
Cool. I'll go ahead and add the fixes for https://openlmis.atlassian.net/browse/OLMIS-5231 here too, then.

jasonrogena committed:
OLMIS-5231: Add Ansible pip Dependencies

Since a bunch of people will be using the defined Ansible playbooks, add
a file with the required pip dependencies. Also add documentation on how
to install the pip and Ansible Galaxy dependencies.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
Fix formatting for NiFi Registry Terraform

Do this by running:
  terraform fmt provision/terraform/modules/nifi-registry
  terraform fmt provision/terraform/reporting/nifi-registry

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
OLMIS-5231: Setup Docker from within uat Terraform

Run the Ansible playbook for setting up Docker and Docker Compose from
within the uat Terraform definitions.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
Fix formatting for uat Terraform

Do this by running:
  terraform fmt provision/terraform/modules/uat
  terraform fmt provision/terraform/uat

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
Rename Terraform Var docker-https-port

Rename the Terraform variable docker-https-port to something more
appropriate (docker-tls-port).

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena commented:
@joshzamor I've added support for setting up Docker and Docker Compose in the uat Terraform deployments. Could you please test those for me 🙏. The Ansible playbook that'll run will create new TLS certs for Docker and update the daemon, so the Jenkins deployments will fail until you update the keys.

jasonrogena committed:
OLMIS-5231: Add DeployGroup EC2 Tag to uat Build

Since the Docker dynamic inventory will only run on hosts that have the
EC2 tag DeployGroup with a value of docker-hosts, add it to the uat
deployments.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

joshzamor commented on provision/terraform/modules/openlmis/variables.tf:
```diff
@@ -1,4 +1,39 @@
+  description = "The TCP port The Docker Daemon is listening for TLS traffic"
+}
+
+variable "app-tls-s3-access-key-id" {
```
So if I understand this right, Ansible will use the AWS access key to save the tls files to, right?  Is this needed or can we just use the AWS access key of the one running Terraform?

jasonrogena replied:
Yeah, that's true, though we'd need to find a way to do that. Terraform doesn't expose the access key it uses. One way we could possibly do this is through Terraform [environment variables](https://www.terraform.io/docs/configuration/environment-variables.html):

```sh
export TF_VAR_app-tls-s3-access-key-id=$AWS_ACCESS_KEY_ID
export TF_VAR_app-tls-s3-secret-access-key=$AWS_SECRET_ACCESS_KEY
```

This however assumes `AWS_SECRET_ACCESS_KEY` and `AWS_ACCESS_KEY_ID` have been set (and Terraform isn't using the `~/.aws/credentials` file).

jasonrogena replied:
Added this to the setup instructions.

joshzamor commented:
Thanks for this @jasonrogena.  I've worked through the changes and I believe I have a general understanding of it.  More so on the Terraform side and less on the Ansible side.  I've left a comment but otherwise I'll be attempting to use this to launch a new OpenLMIS demo.

joshzamor committed:
OLMIS-5289, add benin-demo env

joshzamor committed:
OLMIS-5289, start README for how to use Terraform and Ansible for new env

joshzamor commented on provision/terraform/reporting/nifi-registry/terraform.tfvars:
```diff
@@ -0,0 +1,16 @@
+nr-instance-type = "t2.medium"
+nr-volume-size = "200"
+nr-ssh-key = "ona-terraform-tbd"
```
Could we make sure that this AWS key is in our S3 bucket?

jasonrogena replied:
Done. I've changed the key to `openlmis-reporting`. I've copied the private key to the S3 bucket.

joshzamor reviewed (changes requested):
@jasonrogena I was almost able to use this however I hit a snag in Ansible, and I think it's simply my misunderstanding of what I need to do to install the ansible prereqs correctly.  I started `provision/terraform/README.md` and it'd be great if you could add some high-level steps and requirements for making the whole thing work (e.g. do I need to go to `provision/ansible` and follow the steps in the `README.md` there to install Ansible prereqs?

jasonrogena committed:
OLMIS-5086: Use openlmis-reporting SSH Key

Use the openlmis-reporting SSH key (whose private key is backed up in
the aws-instance-keys S3 bucket) in the NiFi Registry deployment.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
OLMIS-5231: Standardise Terraform AWS Cred Vars

Use the same variable names for the Terraform variables to hold the AWS
key ID and AWS secret key across the UAT and NiFi Registry setups. This
allows for users to get the values from the same Terraform environment
variable.

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
OLMIS-5231: Add Terraform Setup Instructions

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

joshzamor commented on provision/ansible/README.md:
````diff
@@ -0,0 +1,16 @@
+
+```sh
+mkdir -p vendor/roles
+ansible-galaxy install -p vendor/roles -r requirements/galaxy.yml
````
@jasonrogena :  When I try to run this on my machine (OS X.11) I'm getting the following error:

```
ansible-galaxy install -p vendor/roles -r requirements/galaxy.yml
Traceback (most recent call last):
  File "/usr/local/bin/ansible-galaxy", line 67, in <module>
    import ansible.constants as C
  File "/Users/Josh-LT/Library/Python/2.7/lib/python/site-packages/ansible/constants.py", line 11, in <module>
    from jinja2 import Template
ImportError: No module named jinja2
```

I ran `sudo pip install jinja2`, which succeeds, however when I run the ansible command again I get the same error.  I'm not using virtualenv which is not the best idea, however I'm wondering if we're missing something.  Perhaps we can have another dev try this with their machine and see if it works?

joshzamor commented:
@jasonrogena :  I'm still having trouble with Ansible on my machine however I'm afraid this PR will get stale here so I'm going to merge it.  Thanks for the great work here, and if you do have someone else try out the Ansible stuff on their machine I'd be glad to know of it.  I'll keep hacking away on my machine to figure out what's wrong.
